plugins/inputs/http_listener_v2/http_listener_v2.go

```diff
@@ -0,0 +1,304 @@
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+  data_format = "influx"
+`
+
+func (h *HTTPListenerV2) SampleConfig() string {
+	return sampleConfig
+}
+
+func (h *HTTPListenerV2) Description() string {
+	return "Generic HTTP write listener"
+}
+
+func (h *HTTPListenerV2) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (h *HTTPListenerV2) SetParser(parser parsers.Parser) {
+	h.Parser = parser
+}
+
+// Start starts the http listener service.
+func (h *HTTPListenerV2) Start(acc telegraf.Accumulator) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if h.MaxBodySize == 0 {
+		h.MaxBodySize = defaultMaxBodySize
+	}
+
+	if h.ReadTimeout.Duration < time.Second {
+		h.ReadTimeout.Duration = time.Second * 10
+	}
+	if h.WriteTimeout.Duration < time.Second {
+		h.WriteTimeout.Duration = time.Second * 10
+	}
+
+	h.acc = acc
+
+	tlsConf, err := h.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	server := &http.Server{
+		Addr:         h.ServiceAddress,
+		Handler:      h,
+		ReadTimeout:  h.ReadTimeout.Duration,
+		WriteTimeout: h.WriteTimeout.Duration,
+		TLSConfig:    tlsConf,
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", h.ServiceAddress, tlsConf)
+	} else {
+		listener, err = net.Listen("tcp", h.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+	h.listener = listener
+	h.Port = listener.Addr().(*net.TCPAddr).Port
+
+	h.wg.Add(1)
+	go func() {
+		defer h.wg.Done()
+		server.Serve(h.listener)
+	}()
+
+	log.Printf("I! Started HTTP listener V2 service on %s\n", h.ServiceAddress)
+
+	return nil
+}
+
+// Stop cleans up all resources
+func (h *HTTPListenerV2) Stop() {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	h.listener.Close()
+	h.wg.Wait()
+
+	log.Println("I! Stopped HTTP listener V2 service on ", h.ServiceAddress)
+}
+
+func (h *HTTPListenerV2) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	if req.URL.Path == h.Path {
+		h.AuthenticateIfSet(h.serveWrite, res, req)
+	} else {
+		h.AuthenticateIfSet(http.NotFound, res, req)
+	}
+}
+
+func (h *HTTPListenerV2) serveWrite(res http.ResponseWriter, req *http.Request) {
+	// Check that the content length is not too large for us to handle.
+	if req.ContentLength > h.MaxBodySize {
+		tooLarge(res)
+		return
+	}
+
+	// Check if the requested HTTP method was specified in config.
+	isAcceptedMethod := false
+	for _, method := range h.Methods {
+		if req.Method == method {
+			isAcceptedMethod = true
+			break
+		}
+	}
+	if !isAcceptedMethod {
+		methodNotAllowed(res)
+		return
+	}
+
+	// Handle gzip request bodies
+	body := req.Body
+	if req.Header.Get("Content-Encoding") == "gzip" {
+		var err error
+		body, err = gzip.NewReader(req.Body)
+		defer body.Close()
+		if err != nil {
+			log.Println("D! " + err.Error())
+			badRequest(res, err.Error())
+			return
+		}
+	}
+
+	// Add +1 for EOF
+	buf := make([]byte, h.MaxBodySize+1)
+	n, err := io.ReadFull(body, buf)
```

Comment:
Use `ioutil.ReadAll` here, less error prone and will avoid allocating a very large MaxBodySize.  The handling of the data will be much easier too, I think there are 3 cases:
1. request too large: return 413
2. any other error: return 500
3. err is nil, continue on to parser and then return 204

Reply:
I am not sure when it should return 500.
```go
body = http.MaxBytesReader(res, body, h.MaxBodySize)
bytes, err := ioutil.ReadAll(body)
```
`err` type is `errors.errorString` and as far as I tested `err.Error()` only returns `http: request body too large`. Should I compare `err.Error()` to `http: request body too large` if `err != nil`? It doesn't seem that robust to me to check errors by string comparison.

Right now I return:
- 413 if I see an error after `ioutil.ReadAll()`
- 400 if parsing fails
- 200 if it parses successfully

Reply:
I think we are pretty much stuck with error string comparison here since the http package doesn't export a type to compare against.  Just reporting 413 on any error from ReadAll would be a reasonable thing to do in this case as well.

400 if parsing fails is the right one, only return 500 if something unexpected happens that shouldn't be possible.  If we ever see a 500 we will know there is a bug in our code.

Reply:
I still couldn't figure out where to return 500. Can you check the code?

Reply:
Looks like it won't be needed, there are no spots where a server error can occur now.

(Technically something else could happen during ReadAll but as we discussed above we can ignore this)